### PR TITLE
build: add yarn to manage package install

### PR DESCRIPTION
**Pull request by WenheLI**
This pr adds `yarn` support for our monorepo (along with lerna). 
After a bit investigation, I found it is possible to get rid of lerna if we adapt to yarn2. However, this also introduces other problems we can discuss later.
For now, the yarn can surely improve the installing process:
| npm | yarn |
| - | - |
| 170s | 76s |

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,14 +23,15 @@
...
+    "preinstall": "yarn workspace @pipcook/boa run install-dep",
+    "postinstall": "node ./tools/init_dev.js -f && rm -rf ./node_modules/midway-bin/node_modules/typescript/ && yarn workspace @pipcook/boa run compile",
     "docsify": "npm run typedoc && http-server -c-1 ./docs",
     "lint": "eslint . --ext .ts --ext .js && lerna run --scope @pipcook/daemon lint",
     "pretest": "npm run lint",
     "test": "sh run_tests.sh",
     "test:pipeline": "sh ./run_pipeline.sh",
     "typedoc": "sh tools/mkdoc.sh",
-    "clean": "lerna run clean",
+    "clean": "lerna run clean --parallel --concurrency 10",
```
Comment (yorkie) [inline package.json:34]: Shall we use yarn workspace here?

File: packages/app/package.json [hunk 2]
```diff
@@ -12,10 +12,11 @@
     "dev-build": "npm run clean && npm run dev-compile",
     "clean": "rm -rf ./dist && rm -rf tsconfig.tsbuildinfo",
     "dev-compile": "esbuild ./src/*.ts ./src/**/*.ts --tsconfig=tsconfig.json --outdir=dist && sh setup.sh",
-    "compile": "tsc -b tsconfig.json && sh setup.sh"
+    "compile": " tsc -v && tsc -b tsconfig.json && sh setup.sh"
   },
   "license": "Apache 2.0",
   "dependencies": {
+    "crypto": "^1.0.1",
```
Comment (yorkie) [inline packages/app/package.json:19]: Why do we add `crypto` package?

File: package.json (hunk 1 above)
Comment (WenheLI) [reply package.json:34]: Yarn workspace does not support parallel mode at this moment, therefore, to speedup those procedures that could run in parallel we should use `lerna`.

Comment (yorkie) [reply package.json:35]: How about using https://www.npmjs.com/package/concurrently?

File: package.json [hunk 3]
```diff
@@ -16,21 +16,23 @@
...
+    "crypto": "^1.0.1",
     "ts-node": "^8.6.2",
     "typedoc": "^0.17.8",
     "typescript": "^3.9.7"
   },
   "scripts": {
     "build": "lerna run compile",
     "dev-build": "lerna run dev-compile --parallel --concurrency 10",
-    "postinstall": "lerna clean --yes && lerna bootstrap && node ./tools/init_dev.js -f",
+    "preinstall": "yarn workspace @pipcook/boa run install-dep",
```
Comment (yorkie) [inline package.json:27]: Could we move this to boa package?

File: package.json [hunk 4]
```diff
@@ -16,21 +16,23 @@
...
+    "postinstall": "node ./tools/init_dev.js -f && rm -rf ./node_modules/midway-bin/node_modules/typescript/ && yarn workspace @pipcook/boa run compile",
```
Comment (yorkie) [inline package.json:28]: Ditto.

File: package.json [hunk 5]
```diff
@@ -16,21 +16,23 @@
     "jasmine": "^3.5.0",
     "jasmine-spec-reporter": "^5.0.2",
     "lerna": "^3.18.4",
+    "crypto": "^1.0.1",
```
Comment (yorkie) [inline package.json:19]: Which is the dependent of crypto?

File: appveyor.yml [hunk 6]
```diff
@@ -3,7 +3,6 @@ environment:
     - nodejs_version: '8'
     - nodejs_version: '10'
     - nodejs_version: '12'
-    - nodejs_version: '14'
```
Comment (yorkie) [inline appveyor.yml:6]: This might be unnecessary to be changed in this PR.

File: .travis.yml [hunk 7]
```diff
@@ -1,10 +1,9 @@
-sudo: false
+
 language: node_js
 node_js:
   - '8'
   - '10'
   - '12'
-  - '14'
```
Comment (yorkie) [inline .travis.yml:7]: This might be unnecessary to be changed in this PR.

**Comment by yorkie** [comment]
And where do we install yarn?

File: packages/app/package.json [hunk 8]
```diff
@@ -12,7 +12,7 @@
     "dev-build": "npm run clean && npm run dev-compile",
     "clean": "rm -rf ./dist && rm -rf tsconfig.tsbuildinfo",
     "dev-compile": "esbuild ./src/*.ts ./src/**/*.ts --tsconfig=tsconfig.json --outdir=dist && sh setup.sh",
-    "compile": "tsc -b tsconfig.json && sh setup.sh"
+    "compile": " tsc -v && tsc -b tsconfig.json && sh setup.sh"
```
Comment (yorkie) [inline packages/app/package.json:15]: This should be reverted.

File: packages/daemon/bootstrap.js [hunk 9]
```diff
@@ -63,6 +63,9 @@ function createPidfileSync(pathname) {
   if (!await pathExists(midwayPathname)) {
     midwayPathname = path.join(__dirname, '../../midway');
   }
+  if (! await pathExists(midwayPathname)) {
```
Comment (yorkie) [inline packages/daemon/bootstrap.js:66]:
```suggestion
  if (!await pathExists(midwayPathname)) {
```

File: package.json (hunk 3 above)
Comment (WenheLI) [reply package.json:27]: `yarn` will not trigger `postinstall` & `preinstall` in the internal packages if no additional packages are downloaded, therefore, we need to put the two scripts at the root `package.json`

File: package.json (hunk 5 above)
Comment (WenheLI) [reply package.json:19]: removed

File: appveyor.yml (hunk 6 above)
Comment (WenheLI) [reply appveyor.yml:6]: Just came into my mind, this file should be removed, since we are using gihub to do CI/CD.

File: .travis.yml (hunk 7 above)
Comment (WenheLI) [reply .travis.yml:7]: Ditto

**Comment by WenheLI** [comment]
> And where do we install yarn?

Yarn is built-in for github action's container.

**Comment by yorkie** [comment]
> > And where do we install yarn?
> 
> Yarn is built-in for github action's container.

This is not only for GitHub action, our contributor need to install.

File: package.json (hunk 3 above)
Comment (yorkie) [reply package.json:26]: This doesn't work with the separate build :(

File: package.json (hunk 1 above)
Comment (WenheLI) [reply package.json:34]: If we plan to totally get rid of `lerna`, we could add this. 
But for now, we still have some functionalities requires `lerna` (publishing & versioning). I think the best practice is to leave them here. And move to `concurrently` as the moment we do not need `lerna` anymore.

**Comment by WenheLI** [comment]
> > > And where do we install yarn?
> > 
> > 
> > Yarn is built-in for github action's container.
> 
> This is not only for GitHub action, our contributor need to install.

We can add a document about `yarn`  installation.

File: package.json (hunk 1 above)
Comment (yorkie) [reply package.json:33]: But the `clean` is related to build script, and it looks a bit weird that we use yarn to build and lerna to clean files which are built by yarn.

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -55,8 +55,13 @@ jobs:
           git config --global user.name 'pipcook'
           git config --global user.email 'queyue.crk@alibaba-inc.com'
           git add .
+<<<<<<< HEAD
```
Comment (yorkie) [inline .github/workflows/release.yml:58]: Conflicts?

**Review by yorkie — CHANGES_REQUESTED** [review]

**Comment by WenheLI** [comment]
Too many conflicts need to be resolved. 
Will open another PR.

**Pull request closed without merging**